### PR TITLE
Igh input validation

### PR DESCRIPTION
**Pull request by globusharris**
Added input validation checks. 

A few notes:
- Covariance has issues marked in it, which I am working on in a separate branch.
- I did a cursory merge of the input validation into the glm statistic but did not bother to do fixes in objective mech that will allow input validation to run (checkRange input validation is run within censorData), since objective mech is not current priority.

### CODE REVIEW

**Comment by globusharris** [comment]
This resolves issues #99, #100, #101

File: R/mechanism.R [hunk 1]
```diff
@@ -12,7 +12,8 @@
 #' @field n Number of observations
 #' @field epsilon Differential privacy parameter
 #' @field delta Differential privacy parameter
-#' @field rng A priori estimate of the variable range
+#' @field rng A priori estimate of the variable range(s).
+#' @field rngFormat One of 'vector' (for single rng input) or 'list' (for multiple rng inputs)
```
Comment (MeganFantes) [inline R/mechanism.R]: Add a couple more words of description here? How many values in the vector?

File: R/mechanism-objective.R [hunk 2]
```diff
@@ -19,7 +19,7 @@ mechanismObjective$methods(
         x <- x[, cols]
 
         # censor & impute missing values
-        x <- censorData(x, .self$varType, .self$rng, .self$bins)
+        x <- censorData(x, .self$varType, .self$rng, .self$bins) # NEEDS fixed to support checkRange
```
Comment (MeganFantes) [inline R/mechanism-objective.R]: commenting here just to flag the existing comment about fixing this to support checkRange()

File: R/mechanism.R (hunk 1 above)
Comment (MeganFantes) [inline R/mechanism.R:16]: Just curious, what is the purpose of this new parameter?

File: R/statistic-covariance.R [hunk 3]
```diff
@@ -30,7 +30,7 @@ coefficientRelease <- function(formula, release, n) {
 #'    intercept should be added prior to evaluating the inner product x'x.
 #' @return The sensitivity of the data frame for which the covariance matrix
 #'   is being calculated.
-covarianceSensitivity <- function(n, rng, intercept) {
+covarianceSensitivity <- function(n, rng, intercept) { # THIS IS INCORRECT
```
Comment (MeganFantes) [inline R/statistic-covariance.R:33]: flagging existing comment

File: R/statistic-covariance.R [hunk 4]
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
 
 dpCovariance$methods(
   initialize = function(mechanism, varType, n, epsilon, columns, rng=NULL, imputeRng=NULL, 
-                        intercept=FALSE, formula=NULL, delta=1e-5) {
+                        intercept=FALSE, formula=NULL) {
+    # NOTE THIS IS CURRENTLY NOTE SECURE DUE TO SENS AND EPSILON NOT DISTRIBUTED PROPERLY
```
Comment (MeganFantes) [inline R/statistic-covariance.R:124]: flagging existing comment about this not being secure

File: R/statistic-covariance.R [hunk 5]
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
...
     .self$name <- 'Differentially private covariance matrix'
-    .self$mechanism <- mechanism
-    .self$varType <- varType
-    .self$n <- checkNValidity(n)
-    .self$epsilon <- epsilon
-    .self$delta <- delta
-    .self$rng <- rng
+    .self$mechanism <- checkMechanism(mechanism, "mechanismLaplace")
+    .self$varType <- checkVariableType(varType, c('integer', 'double', 'numeric', 'logical'))  #NEED TO CHANGE TO ALLOW MULTIPLE VARTYPES
```
Comment (MeganFantes) [inline R/statistic-covariance.R:127]: flagging existing comment

File: R/statistic-covariance.R [hunk 6]
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
...
     .self$name <- 'Differentially private covariance matrix'
-    .self$mechanism <- mechanism
-    .self$varType <- varType
-    .self$n <- checkNValidity(n)
-    .self$epsilon <- epsilon
-    .self$delta <- delta
-    .self$rng <- rng
+    .self$mechanism <- checkMechanism(mechanism, "mechanismLaplace")
+    .self$varType <- checkVariableType(varType, c('integer', 'double', 'numeric', 'logical'))  #NEED TO CHANGE TO ALLOW MULTIPLE VARTYPES  
+    .self$n <- checkN(n)  #NEED TO ADD THING SAYING HOW THIS IS N PER COLUMN
```
Comment (MeganFantes) [inline R/statistic-covariance.R:128]: could this just be a change where we check n per column within the covariance statistic (e.g. looping through the columns), or would the checkN() function need to be changed?

File: R/statistic-covariance.R [hunk 7]
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
...
+    .self$epsilon <- checkEpsilon(epsilon) #NEED TO CHANGE
```
Comment (MeganFantes) [inline R/statistic-covariance.R:129]: change the function or change the use of the function in the statistic?

File: R/statistic-covariance.R [hunk 8]
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
...
-    checkEpsilon(epsilon)
+    .self$formula <- formula
+    .self$intercept <- intercept
+    
+    checkVariableType(class(formula), "formula")
+    checkVariableType(typeof(intercept), "logical")
+    checkVariableType(typeof(columns), "character")
     
-    if (is.null(imputeRng)) {
+    if (is.null(imputeRng)) { # NEED TO ALLOW FOR IMPUTING ON ONLY SOME OF THE RANGES
```
Comment (MeganFantes) [inline R/statistic-covariance.R:142]: flagging existing comment

File: R/mechanism.R (hunk 1 above)
Comment (globusharris) [reply R/mechanism.R:16]: It flags if the range is a single range or a sequence of ranges. For backwards compatibility (which, idk, maybe a bad idea), I allow single ranges to be input as a vector, e.g. `c(0,5)`. However, you have to enter multiple ranges as a list (or matrix, though those get automatically coerced to lists) because it's the only way to allow people to enter `NULL` for a single range within the series of ranges, which is how the rest of the library treats empty things (as opposed to using `NA`s rather than `NULL`s). And in the first case, when you pass rng, rng[1]=the minimal bound of the range, and in the second, rng[1] is both the minima and the maxima of the range, so you have to treat them differently.

tldr; if we want to support single ranges entered as vectors as currently done, and support matrices/lists of ranges for more complex statistics and also support some of those ranges potentially having missing values or being completely empty/null, the easiest way to do this is to additionally pass in a parameter that describes how the range is formatted. 

side note: given the way R treats `NULL`, we maybe should actually be using `NA` in most of the places we use `NULL`, but as a computer scientist that feels uncomfortable to me, since most of the time we really do mean `NULL`, it's just that R gets fussy about that. 

It's not the prettiest but it works. There are better ways to do it but I ended up doing this after exploring several options and I think it's ok. Might refactor eventually but given time crunch I think I won't bother unless someone has strong feelings about it

File: R/statistic-covariance.R (hunk 6 above)
Comment (globusharris) [reply R/statistic-covariance.R:128]: oh, it only actually needs to be one input value for everything since all columns in data frame should be same size (at lease, I've been assuming that, though I suppose we could have that not be true, hmm). my comment was just to remind myself to add a comment in the covariance statistic to make it clear that by n I mean column length not, say, total number of values across all of the columns you want to find covariance of.

File: R/statistic-covariance.R (hunk 7 above)
Comment (globusharris) [reply R/statistic-covariance.R:129]: kinda both, you have to split up global epsilon in covariance statistic. It's not relevant for this PR, relevant to eventual PR on covariance.

File: R/statistic-histogram.R [hunk 9]
```diff
@@ -82,10 +82,10 @@ dpHistogram$methods(
         
         # get the epsilon and accuracy
         if (is.null(epsilon)) {
-            .self$accuracy <- accuracy
+            .self$accuracy <- checkAccuracy(accuracy)
```
Comment (MeganFantes) [inline R/statistic-histogram.R:85]: why does this checkAccuracy() call not include "expectedLength = 1"?

File: R/statistic-histogram.R [hunk 10]
```diff
@@ -82,10 +82,10 @@ dpHistogram$methods(
...
             .self$epsilon <- histogramGetEpsilon(mechanism, accuracy, delta, alpha, .self$sens)
         } else {
-            .self$epsilon <- epsilon
+            .self$epsilon <- checkEpsilon(epsilon)
```
Comment (MeganFantes) [inline R/statistic-histogram.R:88]: why does this checkEpsilon() call not include "expectedLength = 1"?

File: R/statistic-tree.R [hunk 11]
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
...
-        .self$varType <- varType
+        .self$mechanism <- checkMechanism(mechanism, "mechanismLaplace")
+        .self$varType <- checkVariableType(varType, c('numeric', 'integer', 'logical', 'character'))
         .self$variable <- variable
-        .self$n <- checkNValidity(n)
-        .self$rng <- rng
-        .self$gran <- gran
-        .self$alpha <- alpha
+        .self$n <- checkN(n)
+        .self$rng <- checkRange(rng) # CHANGE
```
Comment (MeganFantes) [inline R/statistic-tree.R:181]: flagging existing comment

File: R/statistic-tree.R [hunk 12]
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
...
+        .self$gran <- checkN(gran, emptyOkay=TRUE) #should be positive whole number
+        .self$alpha <- checkNumeric(alpha)
         .self$sens <- 2 * log2(diff(rng) / gran + 1)
+        
+        checkVariableType(variable, "character")
+        
         if (is.null(epsilon)) {
-            .self$accuracy <- accuracy
+            .self$accuracy <- checkAccuracy(accuracy)
```
Comment (MeganFantes) [inline R/statistic-tree.R:189]: why does this not have an "expectedLength = 1" parameter?

File: R/statistic-tree.R [hunk 13]
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
...
             .self$epsilon <- treeGetParameters(accuracy, rng, gran, alpha)
         } else {
-            .self$epsilon <- epsilon
+            .self$epsilon <- checkEpsilon(epsilon)
```
Comment (MeganFantes) [inline R/statistic-tree.R:192]: why does this not have an "expectedLength = 1" parameter?

File: R/statistic-variance.R [hunk 14]
```diff
@@ -43,19 +43,24 @@ dpVariance$methods(
...
+        .self$rngFormat <- 'vector'
+        .self$rng <- checkRange(rng, .self$varType, .self$rngFormat)
+        
         .self$sens <- (n - 1) / n^2 * diff(.self$rng)^2
         
+        checkVariableType(typeof(variable), 'character')
+        
         if (is.null(epsilon)) {
-            .self$accuracy <- accuracy
+            .self$accuracy <- checkAccuracy(accuracy)
```
Comment (MeganFantes) [inline R/statistic-variance.R:78]: expected length parameter?

File: R/statistic-variance.R [hunk 15]
```diff
@@ -43,19 +43,24 @@ dpVariance$methods(
...
             .self$epsilon <- laplaceGetEpsilon(.self$sens, .self$accuracy, alpha)
         } else {
             checkEpsilon(epsilon)
-            .self$epsilon <- epsilon
+            .self$epsilon <- checkEpsilon(epsilon)
```
Comment (MeganFantes) [inline R/statistic-variance.R:82]: expected length parameter?

File: tests/testthat/test-histogram.R [hunk 16]
```diff
@@ -31,19 +31,13 @@ test_that('expect stability mechanism for unknown variable type', {
     
     expect_error(dpHistogram$new(varType='number data', variable="educ", n=nTest, epsilon=epsilonTest, 
                                  nBins=nBinsTest, delta=deltaTest, rng=c(0,16)), 
-                 "Please enter a data type of 'numeric', 'integer', 'logical', or 'character'")
+                 "Variable type number data should be one of numeric, integer, logical, character", fixed=TRUE)
```
Comment (MeganFantes) [inline tests/testthat/test-histogram.R:34]: shouldn't the value passed into varType be passed into the error message? Right now the error message seems to be hard coded to have the current value of varType.

File: R/statistic-histogram.R (hunk 9 above)
Comment (globusharris) [reply R/statistic-histogram.R:85]: because expectedLength defaults to 1.

File: R/statistic-histogram.R (hunk 10 above)
Comment (globusharris) [reply R/statistic-histogram.R:88]: because expectedLength defaults to 1.

File: R/statistic-tree.R (hunk 12 above)
Comment (globusharris) [reply R/statistic-tree.R:189]: because expectedLength defaults to 1.

File: R/statistic-tree.R (hunk 13 above)
Comment (globusharris) [reply R/statistic-tree.R:192]: because expectedLength defaults to 1.

File: R/statistic-variance.R (hunk 14 above)
Comment (globusharris) [reply R/statistic-variance.R:78]: because expectedLength defaults to 1.

File: R/statistic-variance.R (hunk 15 above)
Comment (globusharris) [reply R/statistic-variance.R:82]: because expectedLength defaults to 1.

File: tests/testthat/test-histogram.R (hunk 16 above)
Comment (globusharris) [reply tests/testthat/test-histogram.R:34]: the error message is hardcoded only in the sense that the expected error in this case is hardcoded in the testthat function. But in general, checkVarType does customize the error message; on a different varType or in a different stat it will output differently.

**Comment by globusharris** [comment]
This resolves issue #99, #100.

**Review by ctcovington — COMMENTED** [review]
I'm unable to build this because of an issue in `vignettes/dp-glm.Rmd` (though this is not changed in this PR as far as I can tell). Seems like things work when I delete that file.

**Comment by globusharris** [comment]
> I'm unable to build this because of an issue in `vignettes/dp-glm.Rmd` (though this is not changed in this PR as far as I can tell). Seems like things work when I delete that file.

Oh that actually makes sense, because I left out refactoring the objective mechanism with the changes in checkRange, since there are so many other issues with the objective mechanism that it seemed silly to do so. I hadn't thought about how that would result in the vignette not building, since we've mostly been completely ignoring that vignette. I'll push a work-around so that we can keep the vignette (at least for now, it probably shouldn't be included in any release that doesn't also include the obj mech fix).

File: R/statistic-tree.R [hunk 17]
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
 )
 
 dpTree$methods(
+    # DO NOT USE
```
Comment (Shoeboxam) [inline R/statistic-tree.R:173]: The initialize function seems like a pretty important function not to use?

**Review by Shoeboxam — APPROVED** [review]

File: R/statistic-tree.R (hunk 17 above)
Comment (globusharris) [reply R/statistic-tree.R:173]: Correct. This is just marking that the entire tree statistic as currently implemented in develop, does not work. Those fixes are in a separate branch that I haven't pushed yet.
